File: python/fastapi/auth-system/routes/auth_routes.py
from fastapi import APIRouter, Depends, HTTPException, status
from fastapi.security import OAuth2PasswordRequestForm
from sqlalchemy.orm import Session
from src.database import get_db
from src.schemas import UserCreate, UserResponse, Token
from src.dependencies import get_current_user, require_role
from src.models import User
import services.auth_service as auth_service

router = APIRouter(prefix="/auth", tags=["authentication"])


@router.post("/register", response_model=UserResponse, status_code=status.HTTP_201_CREATED)
def register(user_data: UserCreate, db: Session = Depends(get_db)):
    """Register a new user (JSON body)."""
    existing_user = auth_service.get_user_by_username(db, user_data.username)
    if existing_user:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Username already registered",
        )
    return auth_service.create_user(db, user_data)


@router.post("/login", response_model=Token)
def login(
    form_data: OAuth2PasswordRequestForm = Depends(),
    db: Session = Depends(get_db),
):
    """
    OAuth2 password login – accepts form-encoded data
    (username & password) and returns a JWT access token.
    """
    user = auth_service.authenticate_user(
        db, form_data.username, form_data.password
    )
    if not user:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Incorrect username or password",
        )
    access_token = auth_service.create_access_token(data={"sub": user.username})
    return Token(access_token=access_token, token_type="bearer")


@router.get("/me", response_model=UserResponse)
def read_me(current_user: User = Depends(get_current_user)):
    """Return the currently authenticated user's profile."""
    return current_user


@router.get("/admin-only", dependencies=[Depends(require_role("admin"))])
def admin_endpoint(current_user: User = Depends(get_current_user)):
    """Admin-only endpoint – requires 'admin' role."""
    return {"message": "Welcome admin!", "user": current_user.username}

File: javascript/express/auth-system/src/routes/auth.routes.ts
import { Router, Request, Response } from 'express';
import { AuthService } from '../services/auth.service';
import { authenticate, requireRole } from '../middleware/auth.middleware';

const router = Router();

// POST /auth/register
router.post('/register', (req: Request, res: Response) => {
  const { email, username, password, role } = req.body;
  if (!email || !username || !password) {
    return res.status(400).json({ message: 'Missing fields' });
  }
  try {
    const user = AuthService.createUser(email, username, password, role || 'user');
    // Don't return the hashed password
    const { hashed_password, ...userWithoutPassword } = user;
    return res.status(201).json(userWithoutPassword);
  } catch (err: any) {
    if (err.message.includes('UNIQUE constraint')) {
      return res.status(400).json({ message: 'Username or email already exists' });
    }
    return res.status(500).json({ message: 'Internal server error' });
  }
});

// POST /auth/login
router.post('/login', (req: Request, res: Response) => {
  const { username, password } = req.body;
  if (!username || !password) {
    return res.status(400).json({ message: 'Missing username or password' });
  }
  const user = AuthService.getUserByUsername(username);
  if (!user || !AuthService.comparePassword(password, user.hashed_password)) {
    return res.status(401).json({ message: 'Invalid credentials' });
  }
  const token = AuthService.generateToken(user);
  return res.json({ access_token: token, token_type: 'bearer' });
});

// GET /auth/me
router.get('/me', authenticate, (req: Request, res: Response) => {
  const user = AuthService.getUserById(req.user!.id);
  if (!user) {
    return res.status(404).json({ message: 'User not found' });
  }
  const { hashed_password, ...userWithoutPassword } = user;
  return res.json(userWithoutPassword);
});

// GET /auth/admin-only
router.get('/admin-only', authenticate, requireRole('admin'), (req: Request, res: Response) => {
  return res.json({ message: 'Welcome admin!', user: req.user!.username });
});

export default router;